src/api_manager/rewrite_rule.cc

```diff
@@ -0,0 +1,353 @@
+}
+
+}  // namespace
+
+bool RewriteRule::ValidateRewriteRule(const std::string &rule,
+                                      std::string *error_msg) {
+  backup_regex_malloc_free();
+
+  std::stringstream ss(rule);
+  std::istream_iterator<std::string> begin(ss);
+  std::istream_iterator<std::string> end;
+  std::vector<std::string> parts(begin, end);
+
+  if (parts.size() != 2) {
+    error_msg->assign("Invalid rewrite rule format(pattern replacement): \"" +
+                      rule + "\"");
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  if (strncmp(parts[1].c_str(), "http://", strlen("http://")) == 0 ||
+      strncmp(parts[1].c_str(), "https://", strlen("https://")) == 0) {
+    error_msg->assign(
+        "Replacement starts with either \"http://\" or \"https://\": is not "
+        "allowed: \"" +
+        rule + "\"");
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  pcre *regex_compiled;
+  pcre_extra *regex_extra;
+
+  const char *pcre_error_str;
+  int pcre_error_offset;
+  regex_compiled = pcre_compile(parts[0].c_str(), 0, &pcre_error_str,
+                                &pcre_error_offset, NULL);
+  if (regex_compiled == NULL) {
+    error_msg->assign("Invalid regular expression in the rewrite rule: \"" +
+                      rule + "\"");
+
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  regex_extra = pcre_study(regex_compiled, 0, &pcre_error_str);
+  if (pcre_error_str != NULL) {
+    error_msg->assign("Invalid regular expression in the rewrite rule: \"" +
+                      rule + "\"");
+
+    pcre_free(regex_compiled);
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  pcre_free(regex_compiled);
+
+  if (regex_extra != NULL) {
+#ifdef PCRE_CONFIG_JIT
+    pcre_free_study(regex_extra);
+#else
+    pcre_free(pcreExtra);
+#endif
+  }
+
+  restore_regex_malloc_free();
+  return true;
+}
+
+RewriteRule::RewriteRule(std::string regex, std::string replacement,
+                         ApiManagerEnvInterface *env)
+    : regex_pattern_(regex),
+      regex_compiled_(NULL),
+      regex_extra_(NULL),
+      replacement_(replacement),
+      env_(env) {
+  backup_regex_malloc_free();
+
+  const char *pcre_error_str;
+  int pcre_error_offset;
+
+  regex_compiled_ = pcre_compile(regex_pattern_.c_str(), 0, &pcre_error_str,
+                                 &pcre_error_offset, NULL);
+  if (regex_compiled_ == NULL) {
+    env_->LogError("Invalid rewrite rule: \"" + regex_pattern_ + "\", error: " +
+                   std::string(pcre_error_str));
+    return;
+  }
+
+  regex_extra_ = pcre_study(regex_compiled_, 0, &pcre_error_str);
+  if (pcre_error_str != NULL) {
+    env_->LogError("Invalid rewrite rule: \"" + regex_pattern_ + "\", error: " +
+                   std::string(pcre_error_str));
+
+    pcre_free(regex_compiled_);
+    return;
+  }
+
+  restore_regex_malloc_free();
+
+  std::string segment;
+  ReplacementPartType status = ReplacementPartType::TEXT;
```

Comment:
Like to see unit-test for this portion of code

Reply:
MatchAndReplacementPattern is testing this part.

Added more cases.